### DISCUSSION TITLE
No completions from `cape-file` when parent directory contains a space

### DISCUSSION THREAD

**Discussion by 3L0C** [minad/cape#136] [Q&A]
![image](https://github.com/user-attachments/assets/6e1f5f41-16de-4081-bd76-961080405765)
![image-1](https://github.com/user-attachments/assets/2c03060e-0387-4800-a354-896e18913c8e)

The first picture shows that candidates are found successfully in `/tmp` for example, but the second picture shows that once I complete to `/tmp/corfu testing/` I no longer get any candidates despite there being two files in the directory. In this example, I accepted the candidate from the list and get nothing afterwards. 

Is this the expected behavior or did I mess something up with my settings?

**Comment by minad** [comment]
Well, just don't use directory names with spaces. :)

**Reply by 3L0C** [reply]
That's certainly one option, lol. 

I'm not too familiar with elisp, but when I insert a separator myself, the completion continues like so:

![image](https://github.com/user-attachments/assets/5ac2284b-0ecb-472b-b45b-1a406fac3b8c)

Is there a way to insert a separator into the accepted candidate or another solution along those lines?

**Comment by myshevchuk** [comment]
Using directory names without spaces is not an option in 2025, of course. Fortunately, Emacs has a ~~function~~ variable just for that. The default value does not include space, so just add a space - or really any other special character you are missing - to the string like this:

``` emacs-lisp
(setq thing-at-point-file-name-chars "-@~/[:alnum:]_.${}#%,: ")
```

**Reply by minad** [reply]
> Using directory names without spaces is not an option in 2025, of course.

Of course it is. Anything works better without spaces, like shells, since you avoid escaping. I think the problem is always that spaces in names will interfere too much with spaces as separators. As another example - how many programming languages do you know which allow spaces in identifiers?

> (setq thing-at-point-file-name-chars "-@~/[:alnum:]_.${}#%,: ")

I wonder if or where this creates problems. Maybe it works well enough given that thing-at-point checks if the file exists. You could propose this change on emacs-devel.

**Reply by myshevchuk** [reply]
The only problem with spaces in file names is old software, which causes problems. Clearly, user space software (pun intended) should work seamlessly with spaces in file names because spaces naturally improve readability. 

Personally, for a long time and until recently I have been omitting spaces in file names as much as possible. Hyphens and underscores are great for file names at which one doesn't stare every day like system files, for example. They are really tiresome in the documents directory though. So I started using spaces where it makes sense for me. So far zero compatibility issues both in shell and Emacs. I would still omit spaces in software packages or script names, etc. That's a nice semantical separation from user documents and a tribute to programming conventions. Being verbose and using spaces in document files and being succinct and sticking to lowercase/hyphen/underscores in system and development.

> I wonder if or where this creates problems. Maybe it works well enough given that thing-at-point checks if the file exists. You could propose this change on emacs-devel.

I will test it and give my feedback here if there are any issues. It would probably be nice to make `thing-at-point-file-name-chars` a user option. The value of this variable looks a little bit arbitrary to me with special Bash characters `$#%`.

**Reply by myshevchuk** [reply]
> As another example - how many programming languages do you know which allow spaces in identifiers?

I don't know how many, but Emacs Lisp allows spaces in identifiers.

**Reply by minad** [reply]
Regarding this bug, it seems to work in my setup in an Org buffer. When I enter `file:dir` it offers `file:dir with space/some file`. But after having entered that link, the link `file:dir with space/some file` is not properly underlined and doesn't work (an Org bug?). It works if I add angle brackets `<file:dir with space/some file>`.

> The only problem with spaces in file names is old software, which causes problems. Clearly, user space software (pun intended) should work seamlessly with spaces in file names because spaces naturally improve readability.

Shells and Emacs are old, but I think they support spaces in file names sufficiently well. But only if spaces are escaped or file names are quoted. My point was about this escaping, which causes problems if forgotten and syntactic overhead. Therefore I think avoiding spaces is still better. Similar problems occur with plain URLs in text. If the URL uses spaces or ends with a dot, the URL is often not recognized correctly, e.g., in mail clients. This problem is similar to the broken Org link with spaces from above.

> I don't know how many, but Emacs Lisp allows spaces in identifiers.

Yes it does, all Lisps do, but you have to escape the space. I meant that you cannot use spaces unescaped, not in the natural form. Writing an identifier in Elisp with a space is like writing a directory name with spaces in shell. In both cases quoting or escaping is needed.

~~~
ls "dir with space"
ls dir\ with\ space
identifier\ with\ space
~~~

Anyway, in the end it is a matter of preference or a trade-off. I understand that looking at a listing with directories with spaces looks cleaner. But that's what one pays for with escaping or quoting in other contexts.

**Reply by myshevchuk** [reply]
> Shells and Emacs are old, but I think they support spaces in file names sufficiently well.

Yes, the things look much better than they used to be say 15 years ago. By old software I didn't mean shells or Emacs but rather some legacy parts that haven't got a chance for an update. 

> But only if spaces are escaped or file names are quoted. My point was about this escaping, which causes problems if forgotten and syntactic overhead. Therefore I think avoiding spaces is still better.

There is definitely an overhead but to my opinion, if the filesystem allows it, then the programs should make their best to allow it too. No spaces in filenames is an _unwritten_ convention as far as I understand it.

> Anyway, in the end it is a matter of preference or a trade-off.

This is how it is whether we like it or not.

**Comment by myshevchuk** [comment]
So I came up with a solution that addresses the `thingatpoint` part without touching or patching the library internals:

``` elisp
(defun bounds-of-filename-at-point ()
  "Get bounds of file name at point.
An improved version of `bounds-of-thing-at-point' for getting file
name bounds, which allows for spaces in file names."
  (let* (bounds-of-thing-at-point-provider-alist
         (thing-at-point-file-name-chars "-@~/[:alnum:]_.${}#%,: ")
         (bounds (bounds-of-thing-at-point 'filename))
         (beg (car bounds))
         (end (cdr bounds))
         (file (buffer-substring-no-properties beg end))
         (file-trimmed (string-trim file))
         (beg-true (string-match (regexp-quote file-trimmed) file))
         (beg (+ beg beg-true))
         (end (+ beg (length file-trimmed))))
    (cons beg end)))

(setq bounds-of-thing-at-point-provider-alist
            (append bounds-of-thing-at-point-provider-alist
                    '((filename . bounds-of-filename-at-point))))
```

It works for plain file names which start with `/`, `~/` or `$SOMEVAR`. Unfortunately, it doesn't help when the file name starts with a prefix like `file:`. Because no spaces is hard coded in `cape-file` when looking for the prefix:

```emacs-lisp
...
(prefix (and cape-file-prefix
             (looking-back
              (concat
               (regexp-opt (ensure-list cape-file-prefix) t)
               "[^ \n\t]*")               ; <= no spaces in file names
              (pos-bol))
             (match-end 1)))
...
```

So to allow for spaces, this portion of the regexp should read `"[^\n\t]*"`.

@minad , could you imagine this part `"[^ \n\t]*"` be made into a user option? Something like `cape-file-prefix-excluded-chars` or whatever?

**Reply by minad** [reply]
No, I don't plan to add special support for spaces here since spaces usually lead to problems with completion, given that space usually terminates a completion field. Also as I wrote above, the resulting `file:` link with spaces will be dysfunctional.

EDIT: I just tried the `"[^\n\t]*"` regexp, and as expected it does not work and lead to issues with auto completion. It is always a trade-off and I consider it more important to not break existing behavior.

**Reply by myshevchuk** [reply]
> Also as I wrote above, the resulting `file:` link with spaces will be dysfunctional.

No, it will be not. For example, Hyperbole doesn't have trouble parsing `file:~/some directory/another dir/some file.el` as is. Org-mode indeed cannot parse it and the above should be surrounded with brackets for the link to work properly. But why is that a concern of Cape what Org can and what it can not? Is it somehow affiliated with Org?

Also, auto-completion with the `file:` prefix works perfectly once the regexp in `cape-file` is set to `"[^\n\t]*"` AND the `thing-at-point` bounds are calculated correctly through a custom `filename` bounds provider function as I suggested above. The latter just goes into user config and is not Cape's concern.

![Screenshot 2025-05-11 at 17 06 23](https://github.com/user-attachments/assets/4f552af9-88bd-4b0f-83b4-7d3841b5561b)

See, Thing-At-Point provides a mechanism to override the defaults and achieve the desired result if the built-in definition is not enough. 

This is exactly what I meant about the user option: make `cape-file` a little bit more flexible by providing a mechanism to override the default regexp. No need for any special support for spaces.

```elisp
(defcustom cape-file-filename-regexp "[^ \n\t]*"  ; <= default value the same as hard-coded
  "Regular expression to match a file name")

;; and then in `cape-file`
...
(prefix (and cape-file-prefix
             (looking-back
              (concat
               (regexp-opt (ensure-list cape-file-prefix) t)
               cape-file-filename-regexp
              (pos-bol))
             (match-end 1)))
...
```

In fact, the de-facto filename regexp `"[^ \n\t]*"` used by Cape is already super permissive since it allows for all kinds of characters except spaces. So the request here is to give the user even more freedom to adjust this regexp to their liking.

Anyway, although a user option would be nice, I'm personally OK to privately advise the `cape-file` function or patch it with `el-patch`. On the other hand, an already substantial and growing user base will keep asking the same question all over again.

**Reply by minad** [reply]
> No, it will be not. For example, Hyperbole doesn't have trouble parsing file:~/some directory/another dir/some file.el as is. Org-mode indeed cannot parse it and the above should be surrounded with brackets for the link to work properly. But why is that a concern of Cape what Org can and what it can not? Is it somehow affiliated with Org?

I looked at Org, and `file:` prefix support was mainly motivated by Org, but of course there can be other modes which happen to support spaces better. Did you considering reporting the problem upstream to Org or do you always rely on Hyperbole for links? I find it concerning if links do not work in some common Emacs modes, and Org is indeed my main use case for this.

> Also, auto-completion with the file: prefix works perfectly once the regexp in cape-file is set to "[^\n\t]*" AND the thing-at-point bounds are calculated correctly through a custom filename bounds provider function as I suggested above.

It did not work sufficiently well when I tried it briefly. If space is removed, auto completion will pop up over aggressively even after space (I mean that's desired here), but often the text after a space is normal text and not part of the link. I'll have to play around a bit more - maybe it can be made to work well enough, even without a customization option. For example `cape-file-directory-must-exist` could be enforced in case a space is found.

> See, Thing-At-Point provides a mechanism to override the defaults and achieve the desired result if the built-in definition is not enough.

Sure. But I think if you want to see this behavior, it should be improved upstream such that everyone profits from it.

> In fact, the de-facto filename regexp "[^ \n\t]*" used by Cape is already super permissive since it allows for all kinds of characters except spaces. So the request here is to give the user even more freedom to adjust this regexp to their liking.

True. But Cape respects the common completion field boundaries, and it happens that these are usually determined by space. So space is special in this sense.

**Reply by myshevchuk** [reply]
> Did you considering reporting the problem upstream to Org or do you always rely on Hyperbole for links?

I typically use `org-insert-link` to insert `file:` links, so I get the brackets for free. It works great with Vertico, by the way, many thanks for it too! Typing file names with the aid of `cape-file`autocompletion comes handy in modes other than Org for paths without the `file:` prefix, for example `(setq some-file-var "~/some-dir/some-file.el")`. Hyperbole is then the only means I'm aware of to follow such a link.

> But I think if you want to see this behavior, it should be improved upstream such that everyone profits from it.

I agree. I don't have much experience with Emacs mailing lists. As far as I remember I've submitted a couple of bug reports to Org in the past. The last one was at least 4-5 years ago and now due to the mailing list wall I would have to relearn the process once again.  I understand your reluctancy to introduce a new and rather obscure customization option, which would not work on its own without either additional user config or an upstream fix. I'll see what I can do.

In any case, thank you for the fruitful discussion and another great package.

**Reply by minad** [reply]
> Hyperbole is then the only means I'm aware of to follow such a link.

Embark and Org (org-open-at-point-global) can do it too, but probably not with spaces...

> I understand your reluctancy to introduce a new and rather obscure customization option, which would not work on its own without either additional user config or an upstream fix.

Yes, that's the point.
